### ISSUE TITLE
Support for specifying config file and resource directory

### ISSUE THREAD

**Issue by NexAdn** [uowuo/abaddon#29]
First of all: nice work!
I just found out about abaddon and it looks very nice.

However, while trying to package it properly, I noticed a two major problems:

- The config file is hard-coded to `$(pwd)/abaddon.ini`.
- When executing abaddon, one must execute it with the directories `css` and `res` pointing to the respective source directories (or rather having their contents).

While the current approce surely is comfortable for a developer, it makes packaging the software hard, as a system-wide installation can not allow config or resource files to sit in `/usr/bin`, where the executable has to sit (at the very least, config files must not sit there as they need to be writable by the user who executes the binary). My current solution is this tiny wrapper script, that does the job, but is not a suitable solution, but rather a workaround:

```sh
#!/bin/sh
set -e

ABADDON_CONFIG_DIR="${HOME}/.config/abaddon"
ABADDON_RES_DIR="/usr/share/abaddon"
ABADDON_BIN="/usr/share/abaddon/abaddon"

mkdir -p "${ABADDON_CONFIG_DIR}"
cd "${ABADDON_CONFIG_DIR}"
ln -sf "${ABADDON_RES_DIR}"/css
ln -sf "${ABADDON_RES_DIR}"/res
exec "${ABADDON_BIN}"
```

The easiest solution would probably be to just take compile-time parameters (e.g. via a `config.h.in` and CMake's `configure_file`) to specify default locations for resource directories or implementing a fallback to `/usr/share/abaddon` (which is probably the resource directory appropriate for most Linux distributions).

For a config file, one might implement a mechanism to first check for a config file in `pwd` and then fall back to e.g. `~/.config/abaddon.ini` or `~/.config/abaddon/abaddon.ini`.

To support multiple configuration files, one might also allow passing the config file location via an environment variable or command-line parameter (the latter one is probably the easier to implement). The same goes for the resource directory, but it's rather unnecessary for that one, as one should always use the resources which are packaged together with the installed binary.

**Comment by ouwou** [comment]
i added the `resources` branch. it should pull `css` and `res` from `/usr/share/abaddon` now if present or fallback to the cwd. it should also now pull abaddon.ini from `~/.config/abaddon/abaddon.ini` if it's not in the cwd too. i'll probably add an environment variable for overriding the config location too since i dont want to deal with parsing a command line

as for the `configure_file` idea i'm not really familiar with that side of CMake so that might take a little bit of figuring out. is there any reason why that would be a compile-time thing instead of runtime like the config file? (i guess its more useful to change the config at runtime or something)

lemme know what u think

**Comment by javsanpar** [comment]
What do you think of pulling `css` and `res` from `~/.config/abaddon/abaddon`? This way, different users in the same machine can have different styles.

**Comment by ouwou** [comment]
> 
> 
> What do you think of pulling `css` and `res` from `~/.config/abaddon/abaddon`? This way, different users in the same machine can have different styles.

yeah that can be done. i guess the priority for this then should be the working directory, then that path, and then from `/usr/share`?

**Comment by javsanpar** [comment]
Seems good to me!

**Comment by NexAdn** [comment]
Hi,

> i added the `resources` branch. it should pull `css` and `res` from
> `/usr/share/abaddon` now if present or fallback to the cwd.
> it should also now pull abaddon.ini from
> `~/.config/abaddon/abaddon.ini` if it's not in the cwd too.
Nice! I will test it out soon.

> i'll probably add an environment variable for overriding the config
> location too since i dont want to deal with parsing a command line
Well, in that case it shouldn't be that hard. It can be as simple as
checking args and then just taking argv[1] as the path to the config
file (if specified, of course–that's what you check argc for).

> as for the `configure_file` idea i'm not really familiar with that
> side of CMake so that might take a little bit of figuring out.
That's not very hard either. Just create a config.h.in file like this:

```c
#define RESOURCE_DIR_DEFAULT "@MY_CMAKE_VARIABLE@"
```

and in CMake, insert

```cmake
configure_file(
    ${PROJECT_SOURCE_DIR}/config.h.in
    ${PROJECT_BINARY_DIR}/config.h
)

target_include_directories(...
    ${PROJECT_BINARY_DIR}
)
```

Then, after configure stage, there is a config.h in your build
directory, where every occurence of `@MY_CMAKE_VARIABLE@` is replaced
with the exact value of the CMake variable with the corresponding
name.

> is there any reason why that would be a compile-time thing instead
> of runtime like the config file?
Well, usually the global resources are in a well-known directory
somewhere below /usr (in this case it's /usr/share/abaddon). While
/usr/share/abaddon is the correct path for most Linux distributions, I
can imagine there being other distributions with different paths.

Lastly, there is still the possibility that one might want install to
a different prefix than /usr, so at the very least, the location
should be more like `${CMAKE_INSTALL_PREFIX}/share/abaddon` than just
plain `/usr/share/abaddon`. This allows for installing e.g. to
`/usr/local` if one doesn't want to cause problems with the package
manager when installing e.g. using `ninja install` or `make install`.

Of course, if it's desired to also allow pulling in the sources from
somewhere else than the well-known system resource directories, a
runtime-mechanism could also be used, but the fallback should depend
on a configure-time variable to allow for different install prefixes.

> (i guess its more useful to change the config at runtime or
> something)
Yes, changing the config at runtime is usually more useful. After all,
one might want to have multiple accounts open at once^^

I might have a look into some of that, too, but I'm not familiar with
the code, so I can't promise anything...

**Comment by ouwou** [comment]
>Well, in that case it shouldn't be that hard. It can be as simple as
checking `argc` and then just taking `argv[1]` as the path to the config
file (if specified, of course–that's what you check argc for).

yeah that'd be the obvious way but it'd cause problems further down the line if i ever decide to add more command line options. i already added an environment variable (`ABADDON_CONFIG`) option so i'll just stick with that unless there's some important benefit for adding it to the command line

as for the resource dir stuff i don't know too much about how linux's file hierarchy works but i think i can figure out what you've said here. in that case i guess the hardcoded /usr/share/abaddon (e.g. [here](https://github.com/uowuo/abaddon/blob/0fcd14c089ea8c3cf281200b90cc1745c7fd7d53/platform.cpp#L100)) would be replaced with that preprocessor definition (the default being `/usr/share/abaddon`)?

and when you bring up `CMAKE_INSTALL_PREFIX` does that even apply here? i've never tried making something `ninja/make install`-able but i'd figure that's something i would have to manually add to CMakeLists. or is this something i don't even have to care about if i do the `configure_file` stuff

**Comment by NexAdn** [comment]
Yeah, if command-line-parameters shall be reserved for future use, an environment variable is not a bad idea. On the other hand, one might also use something like getopt(3) or Boost program options for command-line parsing. Just to give you an idea, what you can use for the job.

> in that case i guess the hardcoded /usr/share/abaddon (e.g. here) would be replaced with that preprocessor definition

I am already concerned that you say “**e.g.** here”. Such things should occur only once in the whole source tree–preferably in some header file which contains all the constants or in a header file of the only class using a certain constant. It's a bit of work to do that initially, but it really pays out if you ever need to change something (like now) or if you'd like to have an easy time with unit tests.

But yeah, in that case you would have something like

```c
#define DEFAULT_RESOURCE_DIR "@CMAKE_INSTALL_PREFIX@" "/abaddon"
```

or

```c
#define DEFAULT_RESOURCE_DIR "@ABADDON_SYSTEM_RESOURCE_DIR@"
```
(given you have a CMake variable `ABADDON_SYSTEM_RESOURCE_DIR` which contains the desired resource directory)

does the job. You can theoretically do the same for almost anything which you might want to configure at compile-time (e.g. optional features, OS-dependent code paths, etc.). Just don't over-use `#ifdef` and its friends–that can make the code very error-prone as you theoretically have to test it for every combination of enabled/disabled options.

> i've never tried making something ninja/make install-able but i'd figure that's something i would have to manually add to CMakeLists. or is this something i don't even have to care about if i do the configure_file stuff

Yes and no. When preparing your code for system-wide installation, you start making the code respect the prefix path (like `/usr` in this case) for certain things (e.g. to find the resource directory). But the install itself is just telling CMake which files to install where and with which permissions.

--------
(this part is a bit off-topic)

You might want to have a look in a few resources regarding CMake installs:
* https://cmake.org/cmake/help/latest/command/install.html
* https://stackoverflow.com/questions/48428647/how-to-use-cmake-to-install
* https://www.foonathan.net/2016/03/cmake-install/
* https://stackoverflow.com/questions/6241922/how-to-use-cmake-install-prefix

(most resources deal with installing libraries, but executables work in a similar way–they are even easier to install)

And if you'd like to install files, you can have a look at [this](https://github.com/bazukas/obs-linuxbrowser/blob/267e04e96bab60677cf70a42f812ccbabcd72b55/CMakeLists.txt#L98-L103)–although I recommand using [GNUInstallDirs](https://cmake.org/cmake/help/latest/module/GNUInstallDirs.html) to get the right directories presented as CMake variables.

PS: You might want to move this to a separate issue if you intend to support installation using CMake as there is some more work to do before you can do a proper installation with it...

**Comment by ouwou** [comment]
>I am already concerned that you say “e.g. here”. Such things should occur only once in the whole source tree–

well in my defense it's already only in one place :^) but yeah i know what you mean

i don't want to concern myself with the cmake install stuff right now but i might come back to it in the future

anyways i'll have a go with the configure_file stuff and i take it that should be it then?

**Comment by NexAdn** [comment]
I've just tested the `resources` branch. Loading resources from `/usr/share` and config from `~/.config/abaddon/abaddon.ini` seems to be working fine. However, the following command-line cause abaddon to fail to create a config file:

```sh
$ ABADDON_CONFIG=/home/nex/build/abaddon/build/tmp ./abaddon
```

The message show was:

> The settings file could not be created

**Comment by NexAdn** [comment]
Another one to the suggestion of allowing custom css/resources under `~/.config/abaddon`:

I don't think that's the right place. These files are not really config files, so they shouldn't be part of a config file directory. Same as with `/usr/share`, I'd suggest allowing them to be put under `~/.local/share` (or rather `~/.local/share/abaddon` of course^^). That way, the distinction between config and resources is clear and the local directory structure (`~/.config` for config and `~/.local/share` for resources) reflects the system directory structure (~~`/etc` for config~~ and `/usr/share` for resources). I am really not a fan of putting anything else than config under `~/.config` as some people like to manage their config files using git to synchronize or share them between machines. While could definitely belong into that category, custom icons etc. most certainly don't. It's already bad enough that OBS puts whole binaries (the plugin `.so` files) in their `~/.config` directory and all Electron Apps have the urge to ignore that there is `~/.cache` and rather put their caches inside `~/.config` which makes cleaning up all those hidden directories below `~` a real pain in the ass.

Those directory structures are there for a reason and I am very happy that abaddon starts to follow them in the system/global scope. Now I'd really like to see the same happen in the per-user scope.

Anyway, +1 for supporting custom CSS/resources on a per-user level and thank you for having implemented the new config file detection so quickly!

**Comment by ouwou** [comment]
> However, the following command-line cause abaddon to fail to create a config file:
> 
> ```shell
> $ ABADDON_CONFIG=/home/nex/build/abaddon/build/tmp ./abaddon
> ```
> 
> The message show was:
> 
> > The settings file could not be created

the environment variable expects the path including the filename

i think your explanation about the home folders makes sense, i'll change the path for the resources out

**Comment by ouwou** [comment]
any more input on this? i'd like to merge it to master now since everything seems to be functioning as expected

**Comment by javsanpar** [comment]
It looks good and ready to merge to me.

Thank you for your efforts!
